### ISSUE TITLE
Nargo prove/verify sha256 takes too long

### ISSUE THREAD

**Issue by Syhawk** [noir-lang/noir#724]
# Description

## Aim

Proving and verifying sha256 16 times.

## Expected behavior

Proving and verifying sha256 16 times consumes little time.

## Bug

```
prove time: 2m32.326s
verify time: 2m14.783s
```

# To reproduce

1. use this code
`main.nr`
```
use dep::std;
fn main(x: pub [u8; 64], result: pub u8) {
    let mut digest = std::hash::sha256(x);
    let N = 15;
    for i in 0..N {
        if (i >= 0) {
            digest = std::hash::sha256(x);
        }
    }
    constrain digest[0] == result;
}
```
`Prover.toml`
```
x = [
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00,
    0x00
]
result = 0xf5
```
2. `nargo prove p`
3. `nargo verify p`

# Environment

- [x] Using `nargo`:

  - [x] Binary: [0.1 ](https://github.com/noir-lang/noir/releases/download/nightly/nargo-aarch64-apple-darwin.tar.gz)
  - [ ] Compiled from source: 

- [ ] Using TypeScript:

  - @noir-lang/noir_wasm: 
  - @noir-lang/barretenberg: 
  - @noir-lang/aztec_backend: 

# Additional context
`Macbook Pro (2021 14-inch M1 Max) 16G`

**Comment by vezenovm** [comment]
I see for your environment you are using the binary. This is most likely why the times are slow. When building `nargo` from source I have a proving time around 13s and a verification time around 10s. This is something we are aware of as we need to speed up the execution of our `aztec_wasm_backend`. 

For the time being you could try building `nargo` from source to have faster proving.

**Comment by Syhawk** [comment]
> 

1. When building `nargo` from source using `aztec_backend`, i get the same result as you.
2. And there is another question about `verifying time`, you can increase the value of `N` and try again, why does the `verifying time` increase linearly?

**Comment by Syhawk** [comment]
> I see for your environment you are using the binary. This is most likely why the times are slow. When building `nargo` from source I have a proving time around 13s and a verification time around 10s. This is something we are aware of as we need to speed up the execution of our `aztec_wasm_backend`.
> 
> For the time being you could try building `nargo` from source to have faster proving.

When building `nargo` from source using `aztec_backend`, i get the same result as you.
And there is another question about `verifying time`, you can increase the value of N and try again, why does the `verifying time` increase linearly?

**Comment by vezenovm** [comment]
> And there is another question about `verifying time`, you can increase the value of N and try again, why does the `verifying time` increase linearly?

Good question. This is something we are aware of and fixing. Our VM interface does not have preprocessing right now, thus we are generating the prover and verifier keys on every prove and verify. I am working towards upgrading this VM interface and should hopefully have an update for you soon :)

**Comment by kevaundray** [comment]
@vezenovm what is the status of this issue?

**Comment by vezenovm** [comment]
When building from source the time to prove/verify sha256 is pretty reasonable. On my local Mac intel (which is quite worn from use) I get times around 5 seconds for both `nargo prove p` and `nargo verify p`. When saving the verification key using `nargo compile` the verification time cuts by another second to around 4 seconds. 

Now, the issue is using the wasm version of nargo (which is going to be replaced by bb.js) and this is also before UltraPlonk. bb.js has significant speed improvements of its own as well as UP for sha256. Either way I ran the same circuit using the `plonk_bn254_wasm` feature with the nargo CLI. As expected the wasm version is significantly slower with `nargo prove p` taking about a 1min and 20 seconds and `nargo verify p` taking about 40 seconds. When saving the pkey and vkey, `nargo prove p <circuit_name>` takes under 50 seconds and `nargo verify p <circuit_name>` takes under 15 seconds. These are much more reasonable than the 2-3minute times we used to have, especially with the optimizations coming in `bb.js`. I lean towards closing this issue for now but defer to you on what you think based off of this info @kevaundray
